_posts/2017/2017-16-01-testing-async-interactions.md

```diff
@@ -0,0 +1,246 @@
+---
+layout: post
+title: "Testing asynchronous interactions in distributed systems"
+date: "2017-01-15 12:11:00"
+image: blog-header/testing-async-interactions.jpg
+categories: tech-stories
+authors: ["Mathias"]
+---
+
+In a distributed architecture, we strive to build self-contained services.
+We try hard to make these services as autonomous as possible, but it is often impossible to build a service that can fulfill its tasks without interaction with other services.
+
+When it comes to service interaction we try to use an event-based approach as much as possible.
+This means interacting asynchronously and outside the user interaction where this is applicable.
+
+Going this route you will soon ask yourself how to test such asynchronous interactions.
+[Spring Cloud Contract](https://cloud.spring.io/spring-cloud-contract/) offers great support for such scenarios.
+
+Let's look at the problem using a simple application consisting of two services and see how Spring Cloud Contract can help us to improve testing the asynchronous interaction between them.
+
+## The Sample Application
+
+Our application consists of two services:
+
+- the product service is responsible for managing product data
+- the checkout service offers cart-related capabilities
+
+The checkout service cannot fulfill the task of adding a certain quantity of a product to a cart without product data.
+So the checkout service listens to product events to hold a copy of relevant data of each product.
+
+This approach allows us to satisfy all user requests in the checkout service from its local database which should be a lot faster than having to ask the product service for product data while the user is waiting for a reply.
```

Comment:
Sentence too long. Better two sentences: This approach allows us to satisfy all user requests in the checkout service from its local database. This is a lot faster than having to ask the product service for product data while the user is waiting for a reply.